src/DateScalar.php

```diff
@@ -0,0 +1,125 @@
+<?php declare(strict_types=1);
+
+namespace MLL\GraphQLScalars;
+
+use DateTimeImmutable;
+use DateTimeInterface;
+use Exception;
+use GraphQL\Error\Error;
+use GraphQL\Error\InvariantViolation;
+use GraphQL\Language\AST\StringValueNode;
+use GraphQL\Type\Definition\ScalarType;
+use GraphQL\Utils\Utils;
+use GraphQL\Utils\Utils as GraphQLUtils;
+use function is_string;
+use function Safe\preg_match;
+use function Safe\substr;
+
+abstract class DateScalar extends ScalarType
+{
+    public function serialize($value): string
+    {
+        if (! $value instanceof DateTimeInterface) {
+            $value = $this->tryParsingDate($value, InvariantViolation::class);
+        }
+
+        return $value->format(static::outputFormat());
+    }
+
+    public function parseValue($value): DateTimeInterface
+    {
+        return $this->tryParsingDate($value, Error::class);
+    }
+
+    public function parseLiteral($valueNode, ?array $variables = null): DateTimeInterface
+    {
+        if (! $valueNode instanceof StringValueNode) {
+            throw new Error(
+                "Query error: Can only parse strings, got {$valueNode->kind}",
+                $valueNode
+            );
+        }
+
+        return $this->tryParsingDate($valueNode->value, Error::class);
+    }
+
+    /**
+     * @template T of Error|InvariantViolation
+     *
+     * @param mixed $value Any value that might be a Date
+     * @param class-string<T> $exceptionClass
+     *
+     * @throws T
+     */
+    protected function tryParsingDate($value, string $exceptionClass): DateTimeInterface
+    {
+        if (is_string($value)) {
+            if (1 !== preg_match(static::regex(), $value, $matches)) {
+                $regex = static::regex();
+                throw new $exceptionClass("Value \"${value}\" does not match \"{$regex}\". Make sure it's ISO 8601 compliant ");
+            }
+
+            if (! $this->validateDate($matches['date'])) {
+                throw new $exceptionClass(
+                    Utils::printSafeJson(
```

Comment:
This will result in a string like `"\"Expected input value...`. You only need to call this upon `$value` before inlining it into the message.